clients/java/src/main/java/io/camunda/zeebe/client/impl/command/ModifyProcessInstanceCommandImpl.java

```diff
@@ -0,0 +1,221 @@
+import io.camunda.zeebe.gateway.protocol.GatewayOuterClass.ModifyProcessInstanceRequest;
+import io.camunda.zeebe.gateway.protocol.GatewayOuterClass.ModifyProcessInstanceRequest.ActivateInstruction;
+import io.camunda.zeebe.gateway.protocol.GatewayOuterClass.ModifyProcessInstanceRequest.TerminateInstruction;
+import io.camunda.zeebe.gateway.protocol.GatewayOuterClass.ModifyProcessInstanceRequest.VariableInstruction;
+import io.grpc.stub.StreamObserver;
+import java.io.InputStream;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+
+public class ModifyProcessInstanceCommandImpl
+    implements ModifyProcessInstanceCommandStep1, ModifyProcessInstanceCommandStep2 {
+
+  private static final String EMPTY_SCOPE_ID = "";
+  private static final long EMPTY_ANCESTOR_KEY = -1L;
+  private final ModifyProcessInstanceRequest.Builder requestBuilder =
+      ModifyProcessInstanceRequest.newBuilder();
+  private final JsonMapper jsonMapper;
+  private final GatewayStub asyncStub;
+  private final Predicate<Throwable> retryPredicate;
+  private ActivateInstruction latestActivateInstruction;
+  private Duration requestTimeout;
+
+  public ModifyProcessInstanceCommandImpl(
+      final long processInstanceKey,
+      final JsonMapper jsonMapper,
+      final GatewayStub asyncStub,
+      final Duration requestTimeout,
+      final Predicate<Throwable> retryPredicate) {
+    requestBuilder.setProcessInstanceKey(processInstanceKey);
+    this.jsonMapper = jsonMapper;
+    this.asyncStub = asyncStub;
+    this.requestTimeout = requestTimeout;
+    this.retryPredicate = retryPredicate;
+  }
+
+  @Override
+  public ModifyProcessInstanceCommandStep2 activateElement(final String elementId) {
+    return activateElement(elementId, EMPTY_ANCESTOR_KEY);
+  }
+
+  @Override
+  public ModifyProcessInstanceCommandStep2 activateElement(
+      final String elementId, final long ancestorElementInstanceKey) {
+    return addActivateInstruction(elementId, ancestorElementInstanceKey);
+  }
+
+  @Override
+  public ModifyProcessInstanceCommandStep3 terminateElement(final long elementInstanceKey) {
+    requestBuilder.addTerminateInstructions(
+        TerminateInstruction.newBuilder().setElementInstanceKey(elementInstanceKey).build());
+    return this;
+  }
+
+  private ModifyProcessInstanceCommandStep2 addActivateInstruction(
+      final String elementId, final long ancestorElementInstanceKey) {
+    final ActivateInstruction activateInstruction =
+        ActivateInstruction.newBuilder()
+            .setElementId(elementId)
+            .setAncestorElementInstanceKey(ancestorElementInstanceKey)
+            .build();
+    latestActivateInstruction = activateInstruction;
+    requestBuilder.addActivateInstructions(activateInstruction);
+    return this;
+  }
+
+  @Override
+  public ModifyProcessInstanceCommandStep1 and() {
+    return this;
+  }
```

Comment:
🔧 A bit of a nitpick, but I think this should reset the latestActivateInstruction, otherwise just casting it allows to call `withVariables` again.

Reply:
It doesn't because it returns a `ModifyProcessInstanceCommandStep1`, which only contains the `activateElement()` and `terminateElement()` methods 🙂